File: ios/Classes/ble/models/BleDevice.swift
//
//  BleDevice.swift
//  flutter_ezw_ble
//
//  Created by Whiskee on 2025/1/3.
//

/**
 * iOS 扫描结果中的 BLE 设备信息。
 *
 * 该模型只描述扫描展示身份，不持有 CBPeripheral；原生内部使用 `(BleDevice, CBPeripheral)`
 * 组合缓存，Flutter 侧只接收这个可序列化模型。
 */
struct BleDevice: Codable {
    /// 设备命中的 BleConfig 名称。
    let belongConfig: String
    /// CoreBluetooth 广播名。
    let name: String
    /// CoreBluetooth peripheral identifier。
    let uuid: String
    /// 业务序列号，用于左右腿/多设备聚合。
    let sn: String
    /// 按业务广播规则解析出的 MAC，iOS 无法直接读取真实 BLE MAC。
    let mac: String
    /// 最近一次扫描回调中的 RSSI。
    let rssi: Int
}

File: lib/core/models/ble_device.dart
import 'dart:io';

import 'package:flutter_ezw_ble/core/models/ble_connect_state.dart';
import 'package:flutter_ezw_ble/core/models/ble_device_hardware.dart';
import 'package:flutter_ezw_ble/core/tools/connect_state_converter.dart';
import 'package:json_annotation/json_annotation.dart';

part 'ble_device.g.dart';

@JsonSerializable()
class BleDevice {
  final String belongConfig;
  //  iOS为UUID，Android为MAC地址
  String uuid;
  String name;
  String sn;
  int rssi;
  //  MAC地址，Android MAC地址等于UUID
  String mac;

  //  连接状态
  @ConnectStateListConverter()
  BleConnectState connectState;
  //  硬件信息
  @JsonKey(includeFromJson: false, includeToJson: false)
  BleDeviceHardware hardware = BleDeviceHardware();

  BleDevice(
    this.belongConfig,
    this.uuid,
    this.name,
    this.sn,
    this.rssi, {
    this.mac = '',
    this.connectState = BleConnectState.none,
  }) {
    if (mac.isEmpty && Platform.isAndroid) {
      mac = uuid;
    }
  }

  factory BleDevice.fromJson(Map<String, dynamic> json) {
    final newBleDevice = _$BleDeviceFromJson(json);
    final hardwareMap = json['hardware'] as Map<String, dynamic>?;
    if (hardwareMap != null) {
      newBleDevice.hardware = BleDeviceHardware.fromJson(hardwareMap);
    }
    return newBleDevice;
  }

  Map<String, dynamic> toJson() {
    final deviceMap = _$BleDeviceToJson(this);
    deviceMap["hardware"] = hardware.toJson();
    return deviceMap;
  }

  BleDevice copy() => BleDevice(
        belongConfig,
        uuid,
        name,
        sn,
        rssi,
        mac: mac,
        connectState: connectState,
      )..hardware = hardware.copy();
}
